rust/tests/round_trip.rs

```diff
@@ -0,0 +1,169 @@
+
+    let messages = mcap::MessageStream::new(&mapped)?;
+
+    let mut tmp = tempfile()?;
+    let mut writer = mcap::Writer::new(BufWriter::new(&mut tmp))?;
+
+    for m in messages {
+        // IRL, we'd add channels, then write messages to known channels,
+        // which skips having to re-hash the channel and its schema each time.
+        // But since here we'd need to do the same anyways...
+        writer.write(&m?)?;
+    }
+    drop(writer);
+
+    let ours = unsafe { Mmap::map(&tmp) }?;
+
+    // Compare the message stream of our MCAP to the reference one.
+    for (theirs, ours) in
+        mcap::MessageStream::new(&mapped)?.zip_eq(mcap::MessageStream::new(&ours)?)
+    {
+        assert_eq!(ours?, theirs?)
+    }
+
+    // We don't use them, but verify the summary offsets.
+    let footer = mcap::read::footer(&ours)?;
+    assert_ne!(footer.summary_offset_start, 0);
+
+    const FOOTER_LEN: usize = 20 + 8 + 1; // 20 bytes + 8 byte len + 1 byte opcode
+    let summary_offset_end = ours.len() - FOOTER_LEN - mcap::MAGIC.len();
+
+    for (i, rec) in mcap::read::LinearReader::sans_magic(
+        &ours[footer.summary_offset_start as usize..summary_offset_end],
+    )
+    .enumerate()
+    {
+        let offset = match rec {
+            Ok(mcap::records::Record::SummaryOffset(sos)) => sos,
+            wut => panic!("Expected summary offset, got {:?}", wut),
+        };
+
+        // We expect these offsets in this (arbitrary) order:
+        match (i, offset.group_opcode) {
+            (0, op::SCHEMA) => (),
+            (1, op::CHANNEL) => (),
+            (2, op::CHUNK_INDEX) => (),
+            (3, op::STATISTICS) => (),
+            _ => panic!("Summary offset {i} was {offset:?}"),
+        };
+
+        // We should be able to read each group from start to finish,
+        // and the records should be the expected type.
+        let group_start = offset.group_start as usize;
+        let group_end = (offset.group_start + offset.group_length) as usize;
+        for group_rec in mcap::read::LinearReader::sans_magic(&ours[group_start..group_end]) {
+            match group_rec {
+                Ok(rec) => assert_eq!(offset.group_opcode, rec.opcode()),
+                wut => panic!("Expected op {}, got {:?}", offset.group_opcode, wut),
+            }
+        }
+    }
+
+    // Verify the summary and its connectivity.
+
+    let summary = mcap::Summary::read(&ours)?.unwrap();
+    assert!(summary.attachment_indexes.is_empty());
+    assert!(summary.metadata_indexes.is_empty());
+
+    // EZ mode: Streamed chunks should match up with a file-level message stream.
+    for (whole, by_chunk) in mcap::MessageStream::new(&ours)?.zip_eq(
+        summary
+            .chunk_indexes
+            .iter()
+            .flat_map(|ci| summary.stream_chunk(&ours, ci).unwrap()),
+    ) {
+        assert_eq!(whole?, by_chunk?);
+    }
+
+    // Hard mode: randomly access every message in the MCAP.
+    // Yes, this is dumb and O(n^2).
```

Comment:
7 minute CI isn't a terrible situation - with this test rust is the long pole, taking the crown from C++ (5 mins). I think we can speed this up a little by writing a custom MCAP with only 10 messages as part of this test instead of using `demo.mcap`. Happy to do that as a separate PR.